packages/flutter/lib/src/painting/image_provider.dart

```diff
@@ -460,11 +476,84 @@ abstract class AssetBundleImageProvider extends ImageProvider<AssetBundleImageKe
   ///
   /// This function is used by [load].
   @protected
-  Future<ui.Codec> _loadAsync(AssetBundleImageKey key) async {
+  Future<ui.Codec> _loadAsync(AssetBundleImageKey key, DecoderCallback decode) async {
     final ByteData data = await key.bundle.load(key.name);
     if (data == null)
       throw 'Unable to read data';
-    return await PaintingBinding.instance.instantiateImageCodec(data.buffer.asUint8List());
+    return await decode(data.buffer.asUint8List());
+  }
+}
+
+class _SizeAwareCacheKey {
+  const _SizeAwareCacheKey(this.providerCacheKey, this.width, this.height);
+
+  final dynamic providerCacheKey;
+
+  final int width;
+
+  final int height;
+
+  @override
+  bool operator ==(dynamic other) {
+    if (other.runtimeType != runtimeType)
+      return false;
+    final _SizeAwareCacheKey typedOther = other;
+    return providerCacheKey == typedOther.providerCacheKey
+        && width == typedOther.width
+        && height == typedOther.height;
+  }
+
+  @override
+  int get hashCode => hashValues(providerCacheKey, width, height);
+}
+
+/// [CustomCacheSizeImage] instructs Flutter to decode the image at the specified
+/// dimensions instead of at native resolution.
+///
+/// This allows finer control of the size of the image in [ImageCache] and is
+/// generally used to reduce the memory footprint of [ImageCache].
+///
+/// The decoded image may still be displayed at resolutions other than the
+/// cached resolution provided here.
+class CustomCacheSizeImage extends ImageProvider<_SizeAwareCacheKey> {
```

Comment:
It make me wonder whether we should make a mixim or override ImageProvider to take into account of width and height.
Current implementation will break those who implement their own decoder.However, it will have no effect for updating their decoder to have new width and height arguments if they don't use this delegator class.
I feel the breakage experience should be. It should only break them if they opted in for width and height resize.